<div style="display: flex; align-items: center; gap: 20px; font-family: Arial, sans-serif;">
    <div style="margin-right: 30px;">
        <img src="../Sources/Logo.png" width="300" style="border-radius: 10px;"/>
    </div>
    <div style="line-height: 1.5;">
        <strong style="font-size: 18px; display: block; margin-bottom: 5px;">
            <a href='https://t.me/M4m4ev_Alex' style="text-decoration: none; color: #1a73e8;">
                Репетитор по информатике | Мамаев Алексей
            </a>
        </strong>
        <span style="font-size: 16px; color: #333;">
            Разбор <a href='https://euler.jakumo.org/problem/16-summa-cifr-stepeni' style="text-decoration: none; color: #1a73e8;">
            16-й задачи</a> проекта Эйлера
        </span>
        <span style="font-size: 16px; color: #333;">
            <br><br><br><br><br><br>ТГК: <a href='https://t.me/M4m4ev_Alex' style="text-decoration: none; color: #1a73e8;">@M4m4ev_Alex</a><br>
        </span>
    </div>
</div>

## Задача 16

Число
$$
2^{15} = 32768
$$

Сумма его цифр равна:
$$
3 + 2 + 7 + 6 + 8 = 26
$$

Необходимо определить сумму цифр числа
$$
2^{1000}
$$

Ответ: 1366

## 1) Последовательный подсчёт цифр

Самый простой способ решения — вычислить число $2^{1000}$, преобразовать его в строку и последовательно просуммировать все цифры.

---

### Алгоритм

1. Вычислить значение `2 ** 1000`.
2. Преобразовать число в строку.
3. Последовательно пройти по всем символам строки.
4. Преобразовать каждый символ в число и прибавить его к общей сумме.
5. Вернуть полученный результат.

---

### Особенности

- решение очень простое и легко читается;
- используется преобразование числа в строку;
- выполняется всего один проход по всем цифрам числа.

---

### Сложность

Пусть `k` — количество цифр в числе.

- **Время:** `O(k)`.
- **Дополнительная память:** `O(k)` — на хранение строкового представления числа.

In [1]:
from time import perf_counter # измерение времени выполнения алгоритмов

In [2]:
def get_result_1() -> int:
    c = 0
    for i in str(2 ** 1_000):
        c += int(i)
    return c


start = perf_counter()
print(get_result_1())
end = perf_counter()

print(f'{end - start:.5f}')

1366
0.00010


## 2) Арифметический подсчёт цифр

В этом решении число не преобразуется в строку. Сумма цифр вычисляется с помощью арифметических операций.

---

### Алгоритм

1. Вычислить значение `2 ** 1000`.
2. Пока число не станет равно нулю:
   - получить последнюю цифру с помощью операции `% 10`;
   - добавить её к общей сумме;
   - удалить последнюю цифру с помощью целочисленного деления `// 10`.
3. Вернуть полученную сумму.

---

### Особенности

- не используется преобразование числа в строку;
- решение основано только на арифметических операциях;
- за одну итерацию обрабатывается одна цифра числа.

---

### Сложность

Пусть `k` — количество цифр в числе.

- **Время:** `O(k)`.
- **Дополнительная память:** `O(1)`.

In [3]:
def get_result_2() -> int:
    c = 0
    n = 2 ** 1_000
    while n > 0:
        c += n % 10
        n //= 10
    return c


start = perf_counter()
print(get_result_2())
end = perf_counter()

print(f'{end - start:.5f}')

1366
0.00017


## 3) Использование встроенных функций

Наиболее компактное решение можно получить, используя встроенные функции Python.

---

### Алгоритм

1. Вычислить значение `2 ** 1000`.
2. Преобразовать число в строку.
3. С помощью `map` преобразовать каждый символ в целое число.
4. Просуммировать все цифры функцией `sum`.

---

### Особенности

- используется минимальное количество кода;
- решение сочетает функции `sum` и `map`;
- хорошо демонстрирует возможности стандартной библиотеки Python.

---

### Сложность

Пусть `k` — количество цифр в числе.

- **Время:** `O(k)`.
- **Дополнительная память:** `O(k)`.

In [4]:
def get_result_3() -> int:
    return sum(map(int, str(2 ** 1_000)))


start = perf_counter()
print(get_result_3())
end = perf_counter()

print(f'{end - start:.5f}')

1366
0.00009


## Сравнение времени выполнения

Для сравнения производительности используется функция `timeit()`, предназначенная специально для измерения времени выполнения небольших участков кода.

Каждый алгоритм запускается **10_000 раз**, после чего общее время делится на количество запусков. Таким образом получается **среднее время выполнения одного запуска**.

In [5]:
from timeit import timeit

funcs = (
    get_result_1,
    get_result_2,
    get_result_3,
)

for i, func in enumerate(funcs, start=1):
    tm = timeit(lambda: func(), number=10_000) / 10_000
    print(f'Алгоритм {i}: {tm:.10f}')

Алгоритм 1: 0.0000258696
Алгоритм 2: 0.0000907021
Алгоритм 3: 0.0000159696


## Вывод

В данной задаче требовалось вычислить сумму цифр числа $2^{1000}$.

Были рассмотрены три различных способа решения. Первый выполняет последовательный подсчёт цифр после преобразования числа в строку. Второй использует только арифметические операции, последовательно извлекая цифры числа. Третий сочетает встроенные функции `sum` и `map`, благодаря чему получается наиболее компактное решение.

Все три алгоритма имеют линейную сложность относительно количества цифр числа. Однако результаты измерений показывают, что использование встроенных функций Python позволяет получить наилучшее время выполнения:

- **Алгоритм 1:** `0.0000258696` с;
- **Алгоритм 2:** `0.0000907021` с;
- **Алгоритм 3:** `0.0000159696` с.

Таким образом, наиболее удачным является **третье решение**. Оно не только самое короткое и наиболее выразительное, но и показывает лучшее время выполнения за счёт эффективной реализации встроенных функций Python.